plugins/widget/plugin.js

```diff
@@ -3426,12 +3460,134 @@
 		// Cache mouse position so it is re-used in events buffer.
 		listeners.push( editable.on( 'mousemove', function( evt ) {
 			y = evt.data.$.clientY;
+			x = evt.data.$.clientX;
+			target = evt.data.getTarget();
 			buffer.input();
 		} ) );
 
 		// Fire drag start as it happens during the native D&D.
 		editor.fire( 'dragstart', { target: evt.sender } );
 
+		function storeFakeParagraph( fakeParagraph ) {
+			if ( fakeParagraph ) {
+				finder.store( fakeParagraph, Number( fakeParagraph.data( 'cke-fake-paragraph' ) ) );
+			}
+		}
+
+		function addFakeParagraphs( cells ) {
+			CKEDITOR.tools.array.forEach( cells, function( cell ) {
+				addFakeParagraph( cell, true );
+				addFakeParagraph( cell );
+			} );
+		}
+
+		function addFakeParagraph( cell, before ) {
+			var element = before ? cell.getFirst() : cell.getLast();
+
+			if ( !element ) {
+				return null;
+			}
+
+			var isText = element.type === CKEDITOR.NODE_TEXT,
+				isInline = element.is && element.is( CKEDITOR.dtd.$inline ),
+				isBr = element.getName && element.getName() === 'br',
+				fakeParagraph;
+
+			if ( !isText && !isInline || isBr ) {
+				return null;
+			}
+
+			fakeParagraph = CKEDITOR.dom.element.createFromHtml( fakeParagraphHtml );
+			fakeParagraph.data( 'cke-fake-paragraph', before ? CKEDITOR.LINEUTILS_BEFORE : CKEDITOR.LINEUTILS_AFTER );
+			cell.append( fakeParagraph, before );
+			fakeParagraphs.push( fakeParagraph );
+
+			return fakeParagraph;
+		}
+
+		/**
+		 * Sorts relations by actual distance.
+		 *
+		 * @param {Object} relations {@link CKEDITOR.plugins.lineutils.finder#relations}
+		 * @param {Object} locations {@link CKEDITOR.plugins.lineutils.locator}
+		 * @param {Object} coordinates
+		 * @param {Number} coordinates.x horizontal coordinate
+		 * @param {Number} coordinates.y vertical coordinate
+		 * @returns [Array] Sorted array representation of relations {@link CKEDITOR.plugins.lineutils.finder#relations}
+		 */
+		function sortByDistance( relations, locations, coordinates ) {
```

Comment:
This function is a bit complicated (or I'm just scared of math). Maybe you could refactor it a bit or add some comments?

Reply:
I'll add some comments.